# puntseq
PuntSeq - Chasing the microbial diversity of Cambridge's freshwater

Clean freshwater is the foundation for the very basis of human health and wellbeing. While most traditional water monitoring approaches test for specific chemicals or pathogens, the direct tracing of aquatic DNA poses a more holistic alternative which has hitherto been underappreciated due to challenges in logistics and investment. Here we present a simple, fast, inexpensive and reliable freshwater diagnostics workflow centred around portable nanopore DNA sequencing. Using  bacterial mock communities, negative controls, and spatiotemporal microbiata from an example river in Cambridge (UK), our study shows how nanopore sequencing can be readily integrated for the assessment of aquatic microbial biodiversity and pollution by depicting fine temporal gradients along the main hydrological axis of an urban-rural interface, and yielding high-resolution pathogen maps that address concerns of public health.a

Here, we provide all additional environmental data, the final classifications (using minimap2, kmer=15)  of all DNA reads (of three sampling dates, April, June, and August 2018, across nine sampling locations) including the rarefied data, all scripts for the downstream analyses (written in R and python, integrated in a markdown file), and an appropriate conda environment.

Using this platform, the user will be able to easily replicate all results presented in this paper [link will be added soon].

See [here](https://github.com/d-j-k/puntseq/tree/master/analysis) for the detailed README of all raw nanopore data pre-processing steps.

See [here](https://www.puntseq.co.uk/) for more updates about the PuntSeq project!

